modules/transformation.py

```diff
@@ -0,0 +1,500 @@
+        'rank': int(metadata['rank']) if metadata.get('rank') else None,
+        'date': '{:%Y_%m_%d}'.format(date),
+        'client': metadata.get('layout', utils.client_name(file_name)).lower(),
+    }
+
+
+class ImportHarJson(beam.DoFn):
+    def process(self, element, **kwargs):
+        file_name, data = element
+        try:
+            page, requests = self.generate_pages(file_name, data)
+            yield beam.pvalue.TaggedOutput('page', page)
+            yield beam.pvalue.TaggedOutput('requests', requests)
+        except Exception:
+            logging.error(
+                f"Unable to unpack HAR, check previous logs for detailed errors. "
+                f"file_name={file_name}, element={element}"
+            )
+
+    @staticmethod
+    def generate_pages(file_name, element):
+        if not element:
+            logging.exception("HAR file read error.")
+            return
+
+        try:
+            har = json.loads(element)
+        except Exception:
+            logging.exception(f"JSON decode failed for: {file_name}")
+            return
+
+        log = har['log']
+        pages = log['pages']
+        if len(pages) == 0:
+            logging.exception(f"No pages found for: {file_name}")
+            return
+
+        status_info = initialize_status_info(file_name, pages[0])
+
+        # STEP 1: Create a partial "page" record so we get a pageid.
+        page = utils.remove_empty_keys(ImportHarJson.import_page(pages[0], status_info).items())
+        if not page:
+            return
+        page_id = page['pageid']
+
+        # STEP 2: Create all the resources & associate them with the pageid.
+        entries, first_url, first_html_url = ImportHarJson.import_entries(log['entries'], page_id, status_info)
+        if not entries:
+            logging.exception(f"import_entries() failed for status_info:{status_info}")
+            return
+        else:
+            # STEP 3: Go back and fill out the rest of the "page" record based on all the resources.
+            agg_stats = ImportHarJson.aggregate_stats(entries, page_id, first_url, first_html_url, status_info)
+            if not agg_stats:
+                logging.exception(f"aggregate_stats() failed for status_info:{status_info}")
+                return
+            else:
+                page.update(agg_stats)
+
+        return page, entries
+
+    # TODO finish porting logic
+    @staticmethod
+    def import_entries(entries, pageid, status_info):
+        requests = []
+        first_url = ''
+        first_html_url = ''
+
+        for entry in entries:
+            ret_request = {
+                'client': status_info['client'],
+                'date': status_info['date'],
+                'pageid': pageid,
+                'crawlid': status_info['crawlid'],
+                # we use this below for expAge calculation
+                'startedDateTime': utils.datetime_to_epoch(entry['startedDateTime']),
```

Comment:
Yes any date parsing function can potentially throw exceptions due to unexpected/malformed input data. This logic was directly copied from the original implementation here
https://github.com/HTTPArchive/legacy.httparchive.org/blob/de08e0c7c94a7da529826f0a4429a9d28b8fdf5e/bulktest/batch_lib.inc#L515

I'm not familiar with PHP error handling, but according to the docs, the previous PHP `strtotime()` function should have produced warnings rather than exceptions, so we could treat this similarly if needed (i.e. produce a null value instead of failing entirely).

Reply:
That might explain some of the differences in max age in https://github.com/HTTPArchive/data-pipeline/issues/43.

Let’s change it to capture any exception and return null like it used to.